contracts/lib/BlockDecoder.sol

```diff
@@ -380,6 +383,40 @@ library BlockDecoder {
         cs.Signature = property(list, 3).toBytes();
     }
 
+    bytes public constant EXTRA_PREFIX = "HeaderNumber";
+    uint8 public constant HEADER_NUMBER_HASH_LEN = 40;
+
+    // the Extra Data with two format:
+    // first: rlp(produce list)
+    // second: "HeaderNumber",HeaderNumber[8byte],HeaderHash[32byte],rlp(produce list)
+    function decodeExtra(bytes memory headerRLPBytes) internal pure returns (uint256[] memory) {
+        RLPReader.RLPItem[] memory list = decodeToHeaderList(headerRLPBytes);
+        RLPReader.RLPItem memory item = list[uint8(HeaderProperty.Extra)];
+        bytes memory extraData = item.toBytes();
+        (bytes memory res, ) = cutExtraPrefix(extraData);
+        return res.toRlpItem().toUintArray();
+    }
+
+    function cutExtraPrefix(bytes memory extraData) internal pure returns (bytes memory, bool) {
+        bytes memory prefix = EXTRA_PREFIX;
+        if (Memory.equals(Memory.dataPtr(prefix), Memory.dataPtr(extraData), EXTRA_PREFIX.length)) {
+            uint256 prefixLen = EXTRA_PREFIX.length + HEADER_NUMBER_HASH_LEN;
+            require(extraData.length > prefixLen, "data too short");
+            uint256 stateLen = extraData.length - prefixLen;
+            bytes memory dataWithNoPrefix = new bytes(stateLen);
+            Memory.copy(Memory.dataPtr(extraData) + prefixLen, Memory.dataPtr(dataWithNoPrefix), stateLen);
```

Comment:
The gas can be saved using 0x4 precompile contract.  See https://ethereum.stackexchange.com/questions/76760/what-s-the-identity-0x4-precompile